using Core.Exceptions;
using System.Net;
using System.Text.Json;
using FluentValidation;

namespace WebApi.Middleware;

public class CustomExceptionHandlerMiddleware(
    RequestDelegate next
)
{
    public async Task Invoke(HttpContext context)
    {
        try
        {
            Console.WriteLine("++++++++++++++++++HandleExceptionAsync++++++++++++++++++++++++++");
            await next(context);
        }
        catch (Exception exception)
        {
            await HandleExceptionAsync(context, exception);
        }
    }

    private static Task HandleExceptionAsync(HttpContext context, Exception exception)
    {



        var code = HttpStatusCode.InternalServerError;
        var result = string.Empty;

        switch (exception)
        {
            case BadRequestException:
                code = HttpStatusCode.BadRequest;
                break;
            case ValidationException validationException:
                code = HttpStatusCode.BadRequest;
                result = JsonSerializer.Serialize(validationException.Errors);
                break;
            case NotFoundException:
                code = HttpStatusCode.NotFound;
                break;
        }
        context.Response.ContentType = "application/json";
        context.Response.StatusCode = (int)code;

        if (result == string.Empty)
        {
            result = JsonSerializer.Serialize(new { error = exception.Message });
        }

        return context.Response.WriteAsync(result);
    }
}
